File: src/libs/ThirdParty.VBSuppliers/Properties/Stuba/StubaSearch.vb
Imports System.Xml
Imports Intuitive
Imports Intuitive.Helpers.Extensions
Imports Intuitive.Net.WebRequests
Imports iVector.Search.Property
Imports ThirdParty.Abstractions
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Lookups
Imports ThirdParty.Abstractions.Models
Imports ThirdParty.Abstractions.Results
Imports ThirdParty.Abstractions.Search.Models
Imports ThirdParty.Abstractions.Search.Support
Imports ThirdParty.VBSuppliers.My.Resources

Public Class StubaSearch
    Inherits ThirdPartyPropertySearchBase

    Private ReadOnly _settings As IStubaSettings

    Private ReadOnly _support As ITPSupport


    Public Overrides Property Source As String = ThirdParties.STUBA

    Public Overrides ReadOnly Property SqlRequest As Boolean = False

    Public Sub New(settings As IStubaSettings, support As ITPSupport)
        _settings = Ensure.IsNotNull(settings, NameOf(settings))
        _support = Ensure.IsNotNull(support, NameOf(support))
    End Sub

    Public Overrides Function BuildSearchRequests(oSearchDetails As SearchDetails, oResortSplits As List(Of ResortSplit), bSaveLogs As Boolean) As List(Of Request)

        Dim oRequests As New List(Of Request)

        Dim requestBodies As New List(Of String)
        For Each resortSplit As ResortSplit In oResortSplits.Where(Function(rs) Not rs.ResortCode.Contains("|"))
            requestBodies.AddRange(BuildRequests(oSearchDetails, resortSplit.ResortCode, resortSplit.Hotels.Select(Function(h) h.TPKey).ToList))
        Next

        Dim resortCodesGroupedByCityCode As ILookup(Of String, String) =
            oResortSplits.Where(Function(rs) rs.ResortCode.Contains("|")) _
                         .ToLookup(Function(rs) rs.ResortCode.Split("|"c)(0),
                                   Function(rs) rs.ResortCode.Split("|"c)(1))

        For Each resortCodeGroup As IGrouping(Of String, String) In resortCodesGroupedByCityCode
            Dim cityCode As String = resortCodeGroup.Key
            Dim regionIdToUse As String
            Dim hotelIDs As IEnumerable(Of String)
            If resortCodeGroup.Count() > 1 Then
                regionIdToUse = cityCode
                hotelIDs = From rs As ResortSplit In oResortSplits.Where(Function(rs) rs.ResortCode.Split("|"c)(0) = regionIdToUse)
                           From hotel As Hotel In rs.Hotels
                           Select hotel.TPKey
            Else
                regionIdToUse = resortCodeGroup.Single()
                hotelIDs = oResortSplits.Single(Function(rs) rs.ResortCode = $"{cityCode}|{regionIdToUse}").Hotels.Select(Function(h) h.TPKey)
            End If
            requestBodies.AddRange(BuildRequests(oSearchDetails, regionIdToUse, hotelIDs.ToList))
        Next

        For Each request As String In requestBodies
            Dim oRequest As New Request
            With oRequest
                .EndPoint = _settings.URL(oSearchDetails)
                .SetRequest(request)
                .Method = eRequestMethod.POST
                .Source = ThirdParties.STUBA
                .TimeoutInSeconds = Me.RequestTimeOutSeconds(oSearchDetails)
                .LogFileName = "Search"
                .CreateLog = bSaveLogs
                .ExtraInfo = New SearchExtraHelper() With {.SearchDetails = oSearchDetails}
                .UseGZip = True
            End With
            oRequests.Add(oRequest)
        Next
        Return oRequests
    End Function

    Private Iterator Function BuildRequests(oSearchDetails As SearchDetails, sResortCode As String, oHotelIDs As List(Of String)) As IEnumerable(Of String)
        Dim iMaxHotels As Integer = _settings.MaxHotelsPerRequest(oSearchDetails)
        If iMaxHotels <= 0 Then
            Yield BuildRequest(oSearchDetails, sResortCode, oHotelIDs)
        Else
            Dim iBatchCount As Integer = Math.Ceiling(oHotelIDs.Count / iMaxHotels).ToSafeInt()
            For i As Integer = 0 To iBatchCount - 1
                Yield BuildRequest(oSearchDetails, sResortCode, oHotelIDs.Skip(i * iMaxHotels).Take(iMaxHotels))
            Next
        End If
    End Function

    Private Function BuildRequest(oSearchDetails As SearchDetails, sResortCode As String, oHotelIDs As IEnumerable(Of String)) As String
        Dim sOrg As String = _settings.Organisation(oSearchDetails)
        Dim sUser As String = _settings.Username(oSearchDetails)
        Dim sPassword As String = _settings.Password(oSearchDetails)
        Dim sVersion As String = _settings.Version(oSearchDetails)
        Dim sCurrencyCode As String = _settings.Currency(oSearchDetails)
        Dim sNationality As String = _settings.Nationality(oSearchDetails)

        Dim request As XElement =
            <AvailabilitySearch>
                <Authority>
                    <Org><%= sOrg %></Org>
                    <User><%= sUser %></User>
                    <Password><%= sPassword %></Password>
                    <Currency><%= sCurrencyCode %></Currency>
                    <Version><%= sVersion %></Version>
                </Authority>
                <RegionId><%= sResortCode %></RegionId>
                <Hotels>
                    <%= From id As String In oHotelIDs
                        Select <Id><%= id %></Id>
                    %>
                </Hotels>
                <HotelStayDetails>
                    <ArrivalDate><%= oSearchDetails.PropertyArrivalDate.ToString("yyyy-MM-dd") %></ArrivalDate>
                    <Nights><%= oSearchDetails.PropertyDuration %></Nights>
                    <Nationality><%= sNationality %></Nationality>
                    <%= From oRoom As RoomDetail In oSearchDetails.RoomDetails
                        Select <Room>
                                   <Guests>
                                       <%= From adult As Integer In Enumerable.Range(0, oRoom.Adults)
                                           Select <Adult/>
                                       %>
                                       <%= From childAge As Integer In oRoom.ChildAges
                                           Select <Child age=<%= childAge %>/>
                                       %>
                                       <%= From infant As Integer In Enumerable.Range(0, oRoom.Infants)
                                           Select <Child age="0"/>
                                       %>
                                   </Guests>
                               </Room>
                    %>
                </HotelStayDetails>
            </AvailabilitySearch>
        Return request.ToString()
    End Function

    Public Overrides Function TransformResponse(oRequests As List(Of Request), searchDetails As SearchDetails, oResortSplits As List(Of ResortSplit)) As TransformedResultCollection

        Dim filteredResponses As New List(Of XmlDocument)
        For Each webRq As Request In oRequests
            Dim responseXml As XDocument = XDocument.Parse(webRq.ResponseString)
            Dim removeNonRefundable As Boolean =
                _settings.ExcludeNonRefundableRates(searchDetails)
            Dim removeUnknownCancellations As Boolean =
                _settings.ExcludeUnknownCancellationPolicys(searchDetails)

            If removeNonRefundable Then
                RemoveNrfResults(responseXml)
            End If
            If removeUnknownCancellations Then
                RemoveUnknownCancResults(responseXml)
            End If
            If searchDetails.Rooms > 1 Then
                RemoveNonCheapestRoomCombinations(responseXml)
            End If

            filteredResponses.Add(responseXml.ToXmlDocument())
        Next
        Dim merged As XmlDocument = XMLFunctions.MergeXMLDocuments("Results", filteredResponses)

        Dim params As New XSLParams
        params.AddParam("RoomCount", searchDetails.Rooms.ToString())
        Return XMLFunctions.XMLStringTransform(merged, StubaSearchRes.StubaSearchXSL, params)
    End Function

    Private Sub RemoveUnknownCancResults(response As XDocument)
        RemoveResultsByCancellationPolicy(response, "Unknown")
    End Sub

    Private Sub RemoveNrfResults(response As XDocument)
        RemoveResultsByCancellationPolicy(response, "NonRefundable")
    End Sub

    Private Sub RemoveResultsByCancellationPolicy(ByVal response As XDocument, ByVal policyStatus As String)
        response.Elements("AvailabilitySearchResult") _
                .Elements("HotelAvailability") _
                .Elements("Result") _
                .Where(Function(r) r.Elements("Room") _
                                    .Elements("CancellationPolicyStatus") _
                                    .Any(Function(cps) cps.Value.ToLower() = policyStatus.ToLower())) _
                .Remove()
    End Sub

    Private Sub RemoveNonCheapestRoomCombinations(response As XDocument)
        For Each hotel As XElement In response.Elements("AvailabilitySearchResult").Elements("HotelAvailability")
            hotel.Elements("Result") _
                 .OrderBy(Function(r) r.Elements("Price") _
                                       .Sum(Function(p) p.Attribute("amt").Value.ToSafeDecimal())) _
                 .Skip(1) _
                 .Remove()
        Next
    End Sub

    Public Overrides Function SearchRestrictions(oSearchDetails As SearchDetails) As Boolean
        Return False
    End Function

    Public Overrides Function ResponseHasExceptions(oRequest As Request) As Boolean
        Return False
    End Function
End Class

File: src/libs/ThirdParty.CSSuppliers/Stuba/StubaSearch.cs
namespace ThirdParty.CSSuppliers.Stuba
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using System.Xml.Linq;
    using Intuitive;
    using Intuitive.Helpers.Extensions;
    using Intuitive.Helpers.Serialization;
    using Intuitive.Net.WebRequests;
    using ThirdParty.Constants;
    using ThirdParty.Interfaces;
    using ThirdParty.Lookups;
    using ThirdParty.Models;
    using ThirdParty.Results;
    using ThirdParty.Search.Models;

    public class StubaSearch : IThirdPartySearch, ISingleSource
    {
        private readonly IStubaSettings _settings;

        private readonly ISerializer _serializer;

        public string Source => ThirdParties.STUBA;

        public StubaSearch(IStubaSettings settings, ISerializer serializer)
        {
            _settings = Ensure.IsNotNull(settings, nameof(settings));
            _serializer = Ensure.IsNotNull(serializer, nameof(serializer));
        }

        public Task<List<Request>> BuildSearchRequestsAsync(SearchDetails searchDetails, List<ResortSplit> resortSplits)
        {
            var requests = new List<Request>();

            var requestBodies = new List<string>();
            
            foreach (var resortSplit in resortSplits.Where(rs => !rs.ResortCode.Contains("|")))
            {
                requestBodies.AddRange(BuildRequests(searchDetails, resortSplit.ResortCode, resortSplit.Hotels.Select(h => h.TPKey).ToList()));
            }

            var resortCodesGroupedByCityCode = resortSplits
                    .Where(rs => rs.ResortCode.Contains("|"))
                    .ToLookup(rs => rs.ResortCode.Split('|')[0], rs => rs.ResortCode.Split('|')[1]);

            foreach (var resortCodeGroup in resortCodesGroupedByCityCode)
            {
                string cityCode = resortCodeGroup.Key;
                string regionIdToUse;
                IEnumerable<string> hotelIDs;

                if (resortCodeGroup.Count() > 1)
                {
                    regionIdToUse = cityCode;
                    hotelIDs = from rs in resortSplits.Where(rs => (rs.ResortCode.Split('|')[0] ?? "") == (regionIdToUse ?? ""))
                               from hotel in rs.Hotels
                               select hotel.TPKey;
                }
                else
                {
                    regionIdToUse = resortCodeGroup.Single();
                    hotelIDs = resortSplits.Single(rs => (rs.ResortCode ?? "") == ($"{cityCode}|{regionIdToUse}" ?? "")).Hotels.Select(h => h.TPKey);
                }

                requestBodies.AddRange(BuildRequests(searchDetails, regionIdToUse, hotelIDs.ToList()));
            }

            foreach (string requestBody in requestBodies)
            {
                var request = new Request
                {
                    EndPoint = _settings.get_URL(searchDetails),
                    Method = eRequestMethod.POST,
                    Source = ThirdParties.STUBA,
                };
                request.SetRequest(requestBody);
                requests.Add(request);
            }

            return Task.FromResult(requests);
        }

        private IEnumerable<string> BuildRequests(SearchDetails searchDetails, string resortCode, List<string> hotelIds)
        {
            int maxHotels = _settings.get_MaxHotelsPerRequest(searchDetails);

            if (maxHotels <= 0)
            {
                yield return BuildRequest(searchDetails, resortCode, hotelIds);
            }
            else
            {
                int batchCount = Math.Ceiling(hotelIds.Count / (double)maxHotels).ToSafeInt();
                
                for (int i = 0; i <= batchCount - 1; i++)
                {
                    yield return BuildRequest(searchDetails, resortCode, hotelIds.Skip(i * maxHotels).Take(maxHotels));
                }
            }
        }

        private string BuildRequest(SearchDetails searchDetails, string resortCode, IEnumerable<string> hotelIds)
        {
            string org = _settings.get_Organisation(searchDetails);
            string user = _settings.get_Username(searchDetails);
            string password = _settings.get_Password(searchDetails);
            string version = _settings.get_Version(searchDetails);
            string currencyCode = _settings.get_Currency(searchDetails);
            string nationality = _settings.get_Nationality(searchDetails);

            var request = new XElement("AvailabilitySearch",
                            new XElement("Authority",
                                new XElement("Org", org),
                                new XElement("User", user),
                                new XElement("Password", password),
                                new XElement("Currency", currencyCode),
                                new XElement("Version", version)
                            ),
                            new XElement("RegionId", resortCode),
                            new XElement("Hotels", from id in hotelIds
                                                   select new XElement("Id", id)),
                            new XElement("HotelStayDetails",
                            new XElement("ArrivalDate", searchDetails.PropertyArrivalDate.ToString("yyyy-MM-dd")),
                            new XElement("Nights", searchDetails.PropertyDuration),
                            new XElement("Nationality", nationality), from oRoom in searchDetails.RoomDetails
                                                                       select new XElement("Room",
                                                                                  new XElement("Guests", from adult in Enumerable.Range(0, oRoom.Adults)
                                                                                                         select new XElement("Adult"), from childAge in oRoom.ChildAges
                                                                                                                                       select new XElement("Child", new XAttribute("age", childAge)), from infant in Enumerable.Range(0, oRoom.Infants)
                                                                                                                                                                                                      select new XElement("Child", new XAttribute("age", "0")))))
                        );
            return request.ToString();
        }

        public TransformedResultCollection TransformResponse(List<Request> requests, SearchDetails searchDetails, List<ResortSplit> resortSplits)
        {
            var transformedCollection = new TransformedResultCollection();

            var results = requests
                .Select(o =>
                    {
                        try
                        {
                            return _serializer.DeSerialize<StubaSearchResponse>(_serializer.CleanXmlNamespaces(o.ResponseXML));
                        }
                        catch
                        {
                            return null;
                        }
                    })
                .Where(o => o is not null)
                .ToList();

            bool removeNonRefundable = _settings.get_ExcludeNonRefundableRates(searchDetails);
            bool removeUnknownCancellations = _settings.get_ExcludeUnknownCancellationPolicys(searchDetails);

            if (removeNonRefundable)
            {
                RemoveNrfResults(results);
            }
            if (removeUnknownCancellations)
            {
                RemoveUnknownCancResults(results);
            }
            if (searchDetails.Rooms > 1)
            {
                RemoveNonCheapestRoomCombinations(results);
            }

            foreach (var stubaResult in results)
            {
                foreach (var hotelAvail in stubaResult.HotelAvailability)
                {
                    foreach (var result in hotelAvail.Result)
                    {
                        for (int index = 0; index < result.Room.Count(); index++)
                        {
                            var room = result.Room[index];

                            var transformedResult = new TransformedResult()
                            {
                                MasterID = hotelAvail.Hotel.id,
                                PropertyRoomBookingID = index + 1,
                                MealBasisCode = room.MealType.code.ToSafeString(),
                                TPKey = hotelAvail.Hotel.id.ToSafeString(),
                                CurrencyCode = stubaResult.Currency,
                                RoomType = room.RoomType.text,
                                Amount = room.Price.amt,
                                TPReference = $"{hotelAvail.hotelQuoteId}|{result.id}",
                                RoomTypeCode = room.RoomType.code.ToSafeString(),
                                NonRefundableRates = room.CancellationPolicyStatus == "NonRefundable"
                            };

                            transformedCollection.TransformedResults.Add(transformedResult);
                        }
                    }
                }
            }

            return transformedCollection;
        }

        private void RemoveUnknownCancResults(IEnumerable<StubaSearchResponse> results)
        {
            RemoveResultsByCancellationPolicy(results, "Unknown");
        }

        private void RemoveNrfResults(IEnumerable<StubaSearchResponse> results)
        {
            RemoveResultsByCancellationPolicy(results, "NonRefundable");
        }

        private void RemoveResultsByCancellationPolicy(IEnumerable<StubaSearchResponse> results, string policyStatus)
        {
            foreach (StubaSearchResponse response in results)
            {
                foreach (HotelAvailability hotelAvail in response.HotelAvailability)
                {
                    hotelAvail.Result = hotelAvail.Result
                        .Where(o => o.Room.Any(r => (r.CancellationPolicyStatus.ToLower() ?? "") == (policyStatus.ToLower() ?? "")))
                        .ToList();
                }
            }
        }

        private void RemoveNonCheapestRoomCombinations(IEnumerable<StubaSearchResponse> results)
        {
            foreach (HotelAvailability hotelAvail in results.SelectMany(o => o.HotelAvailability.Where(r => r.Result.Count > 0)))
            {
                var cheapestResults = new List<Result>();
                cheapestResults.Add(hotelAvail.Result.OrderBy(o => o.Room.Sum(r => r.Price.amt)).FirstOrDefault());
                hotelAvail.Result = cheapestResults;
            }
        }

        public bool SearchRestrictions(SearchDetails searchDetails, string source)
        {
            return false;
        }

        public bool ResponseHasExceptions(Request request)
        {
            return false;
        }
    }
}